addons/mrp/models/mrp_production.py

```diff
@@ -309,6 +310,42 @@ def _compute_locations(self):
             production.location_src_id = production.picking_type_id.default_location_src_id.id or fallback_loc.id
             production.location_dest_id = production.picking_type_id.default_location_dest_id.id or fallback_loc.id
 
+    def _search_components_availability_state(self, operator, value):
+        def get_stock_moves(moves, state):
+            if state == 'available':
+                return moves.filtered(lambda m: m.forecast_availability == m.product_qty and not m.forecast_expected_date)
+            elif state == 'expected':
+                return moves.filtered(lambda m: m.forecast_availability == m.product_qty and m.forecast_expected_date and m.forecast_expected_date <= m.raw_material_production_id.date_start)
+            elif state == 'late':
+                return moves.filtered(lambda m: m.forecast_availability == m.product_qty and m.forecast_expected_date and m.forecast_expected_date > m.raw_material_production_id.date_start)
+            elif state == 'unavailable':
+                return moves.filtered(lambda m: m.forecast_availability < m.product_qty)
+            else:
+                raise UserError(_('Selection not supported'))
+
+        raw_stock_moves = self.env['stock.move'].search([('raw_material_production_id', '!=', False), ('raw_material_production_id.state', 'not in', ('cancel', 'done', 'draft'))])
+        if operator == '!=' and not value:
+            pass
+        elif operator == '=' and not value:
+            raw_stock_moves = self.env['stock.move'].search([('raw_material_production_id', '!=', False), ('raw_material_production_id.state', 'in', ('cancel', 'done', 'draft'))])
+        elif operator == '=':
```

Comment:
To avoid the first `search` when it's not needed + domain on multiple lines for better readability
```suggestion
        if operator == '!=' and not value:
            pass
        elif operator == '=' and not value:
            raw_stock_moves = self.env['stock.move'].search([
                ('raw_material_production_id', '!=', False),
                ('raw_material_production_id.state', 'in', ('cancel', 'done', 'draft'))
            ])
            return [('move_raw_ids', 'in', raw_stock_moves.ids)]

        raw_stock_moves = self.env['stock.move'].search([
            ('raw_material_production_id', '!=', False),
            ('raw_material_production_id.state', 'not in', ('cancel', 'done', 'draft')),
        ])
        if operator == '=':
```